src/library_webgpu.js

```diff
@@ -202,22 +191,28 @@ var LibraryWebGPU = {
         "texture": this.mgrTexture.get(
           {{{ makeGetValue('ptr', C_STRUCTS.WGPUTextureCopyView.texture, '*') }}}),
         "mipLevel": {{{ gpu.makeGetU32('ptr', C_STRUCTS.WGPUTextureCopyView.mipLevel, '*') }}},
-        "arrayLayer": {{{ gpu.makeGetU32('ptr', C_STRUCTS.WGPUTextureCopyView.arrayLayer, '*') }}},
         "origin": WebGPU.makeOrigin3D(ptr + {{{ C_STRUCTS.WGPUTextureCopyView.origin }}}),
       };
     },
 
-    makeBufferCopyView: function(ptr) {
+    makeTextureDataLayout: function(ptr) {
       {{{ gpu.makeCheckDescriptor('ptr') }}}
       return {
-        "buffer": this.mgrBuffer.get(
-          {{{ makeGetValue('ptr', C_STRUCTS.WGPUBufferCopyView.buffer, '*') }}}),
-        "offset": {{{ gpu.makeGetU64('ptr', C_STRUCTS.WGPUBufferCopyView.offset) }}},
-        "bytesPerRow": {{{ gpu.makeGetU32('ptr', C_STRUCTS.WGPUBufferCopyView.bytesPerRow) }}},
-        "rowsPerImage": {{{ gpu.makeGetU32('ptr', C_STRUCTS.WGPUBufferCopyView.rowsPerImage) }}},
+        "offset": {{{ gpu.makeGetU64('ptr', C_STRUCTS.WGPUTextureDataLayout.offset) }}},
+        "bytesPerRow": {{{ gpu.makeGetU32('ptr', C_STRUCTS.WGPUTextureDataLayout.bytesPerRow) }}},
+        "rowsPerImage": {{{ gpu.makeGetU32('ptr', C_STRUCTS.WGPUTextureDataLayout.rowsPerImage) }}},
       };
     },
 
+    makeBufferCopyView: function(ptr) {
+      {{{ gpu.makeCheckDescriptor('ptr') }}}
+      var layoutPtr = ptr + {{{ C_STRUCTS.WGPUBufferCopyView.layout }}};
+      var bufferCopyView = this.makeTextureDataLayout(layoutPtr);
+      bufferCopyView["buffer"] = this.mgrBuffer.get(
+          {{{ makeGetValue('ptr', C_STRUCTS.WGPUBufferCopyView.buffer, '*') }}});
+      return bufferCopyView;
+    },
+
```

Comment:
I can't find the right line in this file... Can we add a `.slice()` of the data for CreateShaderModule? Chrome can't accept SAB there right now. (if buf.buffer is an instance of SAB)

Reply:
Done, but @kripken what's the correct way to detect whether the heap is a SharedArrayBuffer? Just `instanceof SharedArrayBuffer`?

Reply:
You can check the `USE_PTHREADS` setting. So you can do stuff like
```javascript
#if USE_PTHREADS
  .. do pthreads thing ..
#else
  .. do normal thing ..
#endif
```

Reply:
or
```javascript
{{{ USE_PTHREADS ? "foo" : "bar" }}}
```

Reply:
Is it guaranteed that if USE_PTHREADS is set, the heap is always SAB? We don't want to do an extra copy unnecessarily if the heap could be non-shared.

Although this is a deprecated path (that will get deleted) so it's not that important.

Reply:
Yeah, heap will be a SAB when (and only when) `USE_PTHREADS` mode is in effect.